src/os/bluestore/bluestore_types.cc

```diff
@@ -1080,6 +1080,149 @@ bool bluestore_blob_t::release_extents(bool all,
+      anchor_it->length = offset;
+      ++anchor_it;
+      released_disk->emplace_back(p->offset + offset, length);
+      released_length += length;
+      anchor_it = extents.insert(anchor_it, 2, bluestore_pextent_t(EMPTY, length));
+      ++anchor_it;
+      anchor_it->offset = p_offset + offset + length;
+      anchor_it->length = p_length - offset - length;
+    } else {
+      released_disk->emplace_back(p->offset, length);
+      released_length += length;
+      if (empty) {
+        empty->length += length;
+      } else {
+        anchor_it = extents.insert(anchor_it, 1, bluestore_pextent_t(EMPTY, length));
+        ++anchor_it;
+      }
+      anchor_it->offset = p_offset + length;
+      anchor_it->length = p_length - length;
+    }
+  } else {
+    // p->length > offset
+    // offset + length >= p->length
+    if (offset > 0) {
+      //activate hold, put pextent that we need; put new empty
+      ceph_assert(p->is_valid());
+      hold[0].offset = p->offset;
+      hold[0].length = offset;
+      hold[1].offset = EMPTY;
+      hold[1].length = 0;
+      empty = &hold[1];
+      hold_size = 2;
+    } else {
+      // offset == 0
+      if (empty == nullptr) {
+        //we need empty, activate hold
+        hold[0].offset = EMPTY;
+        hold[0].length = 0;
+        empty = &hold[0];
+        hold_size = 1;
+      }
+    }
+    // starts copying remainder
+    if (p->length - offset) {
+      released_disk->emplace_back(p->offset + offset, p->length - offset);
+    }
+    released_length += p->length - offset;
+    empty->length += p->length - offset;
+    rem -= (p->length - offset);
+    ++p;
+    while (rem > 0 && p->length <= rem) {
+      ceph_assert(p->is_valid());
+      released_disk->emplace_back(p->offset, p->length);
+      released_length += p->length;
+      empty->length += p->length;
+      rem -= p->length;
+      ++p;
+    }
+    if (rem > 0) {
+      ceph_assert(p->is_valid());
+      // this we release
+      released_disk->emplace_back(p->offset, rem);
+      released_length += rem;
+      empty->length += rem;
+      // this much remains
+      p->offset = p->offset + rem;
+      p->length = p->length - rem;
+      //no ++p here; we need this modified p remain part of PExtentVector
+    } else {
+      //amazing, clean cut
+      //if the extent here is empty, we try to meld it
+      if (p != end && !p->is_valid()) {
+        empty->length += p->length;
+        ++p;
+      }
+    }
+    // we erase <anchor, p)
+    // and insert hold in this place
+    int32_t insert_element_cnt = hold_size - (p - anchor);
+    auto anchor_it = extents.begin() + (anchor - begin);
+    if (insert_element_cnt != 0) {
+      if (insert_element_cnt > 0) {
+        anchor_it = extents.insert(anchor_it, insert_element_cnt, bluestore_pextent_t(0, 0));
+      } else {
+        anchor_it = extents.erase(anchor_it, anchor_it + (-insert_element_cnt));
+      }
+    }
+    for (uint32_t i = 0; i < hold_size; i++) {
```

Comment:
hint: one could use std::copy if define hold as std::array.

Reply:
I loot at this function and I think it is too clever.
What do you think about refactor to make it 2 phased:
1) find it_begin, offset_begin; it_end, offset_end
2) move to released, cut out KNOWN region
do you feel it might produce clearer code?